python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeInferenceV2Test.java

```diff
@@ -1231,8 +1243,99 @@ def f():
     var candidates = ((UnionType) iType).candidates();
     Assertions.assertThat(candidates)
       .allMatch(FunctionType.class::isInstance)
-      .extracting(PythonType::name)
-      .contains("foo", "bar");
+      .contains(fooType, barType);
+  }
+
+  @Test
+  @Disabled("Need to support collection item type modification")
+  void inferLoopOverModifiedList() {
+    var root = inferTypes("""
+      def f():
+        l = [1,2,"3"]
+        l.append("1")
+        for i in l:
+          i
+      """);
+
+    var iType = TreeUtils.firstChild(root, ForStatement.class::isInstance)
+      .map(ForStatement.class::cast)
+      .map(ForStatement::body)
+      .flatMap(b -> TreeUtils.firstChild(b, ExpressionStatement.class::isInstance))
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(UnionType.class);
+
+    var candidates = ((UnionType) iType).candidates();
+    Assertions.assertThat(candidates)
+      .allMatch(ObjectType.class::isInstance)
+      .extracting(PythonType::unwrappedType)
+      .contains(INT_TYPE, STR_TYPE);
+  }
+
+  @Test
+  @Disabled("Need to support collection item type modification")
+  void inferLoopOverFlowSensitivePopulatedItemsList() {
+    var root = inferTypes("""
+      def f(b):
+        a = 1
+        if (b):
+          a = "1"
+        l = [a]
+        for i in l:
+          i
+      """);
+
+    var iType = TreeUtils.firstChild(root, ForStatement.class::isInstance)
+      .map(ForStatement.class::cast)
+      .map(ForStatement::body)
+      .flatMap(b -> TreeUtils.firstChild(b, ExpressionStatement.class::isInstance))
+      .flatMap(expressionStatement -> TreeUtils.firstChild(expressionStatement, Name.class::isInstance))
+      .map(Name.class::cast)
+      .map(Expression::typeV2)
+      .get();
+
+    Assertions.assertThat(iType).isInstanceOf(UnionType.class);
+
+    var candidates = ((UnionType) iType).candidates();
+    Assertions.assertThat(candidates)
+      .allMatch(ObjectType.class::isInstance)
+      .extracting(PythonType::unwrappedType)
+      .contains(INT_TYPE, STR_TYPE);
+  }
+
+  @Test
+  @Disabled("The item type for loop iteration should be taken from __iter__ method return type")
+  void inferLoopOverCustomIterableVarType() {
```

Comment:
I'm not 100% sure about this, as it involves declared types on top of custom iterable, and we don't know yet how much trust we want to give to declared types.

I'd suggest saying "We should consider the declared return type of the `__iter__` method here" to be less categoric.